docs/proposals-accepted/202304-query-path-tenancy.md

```diff
@@ -0,0 +1,139 @@
+  * Per-tenant settings. For example, having different limits per tenant, which is a common request. 
+  * Enforce presence of one or more tenant labels in queries.
+
+### Audience
+
+<!--
+If this is not clear already, provide the target audience for this change.
+-->
+
+Any team running Thanos in a multi-tenant environment.
+
+## Non-Goals
+
+<!--
+* Move old designs to the new format.
+* Not doing X,Y,Z.
+-->
+
+* Add multi-tenancy to Thanos Ruler. It's not needed. One will still be able to point their Ruler to the multi-tenant Query Frontend, effectively making it multi-tenant. 
+* Implement cross-tenant querying. It poses the question about how to track metrics of a multi-tenant query, probably requiring new logic for query splitting and/or separation of the concepts of query initiator tenant and query target tenants. To exemplify: if tenant A executes a query touching their own data and tenant B data (imagine a label matcher `tenant_id=~"A|B"`), how does Thanos export a request duration metric about this query? This is out of scope and an specific proposal can be created for it.
+
+## How
+
+<!--
+Explain the full overview of the proposed solution. Some guidelines:
+
+* Make it concise and **simple**; put diagrams; be concrete, avoid using “really”, “amazing” and “great” (:
+* How you will test and verify?
+* How you will migrate users, without downtime. How we solve incompatibilities?
+* What open questions are left? (“Known unknowns”)
+-->
+
+* Implement a mechanism to allow incoming requests to specify the tenant being queried using an HTTP header. This applies to all components in the query path. We follow the example of Thanos Receive, which uses the `--receive.tenant-header="THANOS-TENANT"` flag to configure the tenant header, adapting its name to each component.
+* The tenant header value from a given request should travel downstream to all the components being called, so that it can be added to their metrics, traces, and logs without requiring duplicated/extra work to re-parse the query. This also applies to gRPC calls.
+* Implement a command line argument in the Querier component that will indicate with label name that should be used to enforce tenancy. Again, we follow the example of Thanos Receive, which uses the `--receive.tenant-label-name="tenant_id"` flag to achieve this, and adapt its name. The default behavior, when the flag as an empty string as value, is to not identify tenants. 
+  * The label verification and enforcement should be done by reusing prom-label-proxy's [Enforce.EnforceMatchers](https://github.com/prometheus-community/prom-label-proxy/blob/main/injectproxy/enforce.go#L141). There's no reason to (re)implement something specific and special for Thanos.
+  * This behavior should be implemented as part of the base logic of both the HTTP and gRPC query APIs, before the query is handed to the query engine. This allows users managing complex Querier trees can choose where they want this logic to be enabled. 
```

Comment:
What about StoreAPI, which is mentioned in the architecture diagram? Also, maybe worth mentioning how the header propagation flow would work for different architectures? 🙂

Reply:
There's no need to add or change anything at all in the Store API definition. As mentioned there in the text, the HTTP header will be passed even on gRPC requests, which in the end are also HTTP requests. So it'll handle the header just as any other server will: check if it's present and use it as an extra label in its own exported metrics.

Reply:
The behavior of the Store API server doesn't change at all from the point of view of a Store API client. The only component that will check the tenant header value against the tenant label selector is the Querier.